### PR TITLE
Fallback to nn batchnorm on backward-evaluate

### LINKED ISSUE #586: CUDNN batchnorm backprop doesn't work properly in evaluation mode
Gradient changes, though inputs, outputs, and internal state stay same.
```python
cudnn.enabled = True

net = nn.BatchNorm1d(3).cuda()
net.eval()

idat = Variable(torch.rand(4,3).cuda(), requires_grad=True)
grad = torch.rand(4,3).cuda()

# 1st pass
res = net(idat)
res.backward(grad)
grad0 = idat.grad.data.cpu()

# 2nd pass
idat.grad.data.zero_()

res = net(idat)
res.backward(grad)
grad1 = idat.grad.data.cpu()

print (grad1 == grad0).all()
```
Even when I run forward in evaluation mode and backward in training I still get the gradient changed.
It works fine with cudnn disabled.

### PR DESCRIPTION
**Pull request by szagoruyko**
fix https://github.com/pytorch/pytorch/issues/586

### CODE REVIEW

**Comment by colesbury** [comment]
Looks good. Can you add a test to `test_nn.py`?

**Comment by soumith** [comment]
merging it as it's a  critical fix, hoping that tests will come later :)
